src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,357 @@
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal interface ITypeImportCompletionService : IWorkspaceService
+    {
+        ImmutableArray<TypeImportCompletionItem> GetAccessibleTopLevelTypesFromPEReference(
+            Solution solution,
+            Compilation compilation,
+            PortableExecutableReference peReference,
+            ImmutableHashSet<string> excludedNamespaces,
+            CancellationToken cancellationToken);
+
+        Task<ImmutableArray<TypeImportCompletionItem>> GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+            Solution solution,
+            Compilation compilation,
+            CompilationReference compilationReference,
+            ImmutableHashSet<string> excludedNamespaces,
+            CancellationToken cancellationToken);
+
+        Task<ImmutableArray<TypeImportCompletionItem>> GetAccessibleTopLevelTypesFromProjectAsync(
+            Project project,
+            ImmutableHashSet<string> excludedNamespaces,
+            CancellationToken cancellationToken);
+    }
+
+    [ExportWorkspaceServiceFactory(typeof(ITypeImportCompletionService), ServiceLayer.Editor), Shared]
+    internal sealed class TypeImportCompletionService : IWorkspaceServiceFactory
+    {
+        private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache
+            = new ConcurrentDictionary<string, ReferenceCacheEntry>();
+
+        private readonly ConcurrentDictionary<ProjectId, ReferenceCacheEntry> _projectItemsCache
+            = new ConcurrentDictionary<ProjectId, ReferenceCacheEntry>();
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            var cacheService = workspaceServices.GetService<IWorkspaceCacheService>();
+            if (cacheService != null)
+            {
+                cacheService.CacheFlushRequested += OnCacheFlushRequested;
+            }
+
+            return new Service(_peItemsCache, _projectItemsCache);
+        }
+
+        private void OnCacheFlushRequested(object sender, EventArgs e)
+        {
+            _peItemsCache.Clear();
+            _projectItemsCache.Clear();
+        }
+
+        private class Service : ITypeImportCompletionService
+        {
+            private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache;
+            private readonly ConcurrentDictionary<ProjectId, ReferenceCacheEntry> _projectItemsCache;
+
+            public Service(ConcurrentDictionary<string, ReferenceCacheEntry> peReferenceCache, ConcurrentDictionary<ProjectId, ReferenceCacheEntry> projectReferenceCache)
+            {
+                _peItemsCache = peReferenceCache;
+                _projectItemsCache = projectReferenceCache;
+            }
+
+            public async Task<ImmutableArray<TypeImportCompletionItem>> GetAccessibleTopLevelTypesFromProjectAsync(
```

Comment:
are these just source types? if so, it would be good to mentione.

Reply:
Added comments and now it throws if project doesn't support compilation